nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/layers/recurrent/outputs/LSTMLayerOutputs.java

```diff
@@ -2,178 +2,132 @@
 
-    private RnnDataFormat dataFormat;
-
     /**
-     * Output - input modulation gate activations.
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
+     * The LSTM layer data format ({@link LSTMDataFormat}.
      */
-    private SDVariable i;
+    private LSTMDataFormat dataFormat;
 
-    /**
-     * Activations, cell state (pre tanh).
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
-     */
-    private SDVariable c;
 
     /**
-     * Output - forget gate activations.
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
+     * output h:
+     * [sL, bS, nOut]    when directionMode <= 2 && dataFormat == 0
+     * [bS, sL, nOut]    when directionMode <= 2 && dataFormat == 1
+     * [bS, nOut, sL]    when directionMode <= 2 && dataFormat == 2
+     * [sL, bS, 2*nOut]  when directionMode == 3 && dataFormat == 0
+     * [bS, sL, 2*nOut]  when directionMode == 3 && dataFormat == 1
+     * [bS, 2*nOut, sL]  when directionMode == 3 && dataFormat == 2
+     * [sL, 2, bS, nOut] when directionMode == 4 && dataFormat == 3
+     * numbers mean index in corresponding enums {@link LSTMDataFormat} and {@link LSTMDirectionMode}
      */
-    private SDVariable f;
+    private SDVariable h;
 
-    /**
-     * Output - output gate activations.
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
-     */
-    private SDVariable o;
 
     /**
-     * Output - input gate activations.
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
+     * cell state at last step cL:
+     * [bS, nOut]   when directionMode FWD or BWD
+     * 2, bS, nOut] when directionMode  BIDIR_SUM, BIDIR_CONCAT or BIDIR_EXTRA_DIM
      */
-    private SDVariable z;
+    private SDVariable cL;
 
     /**
-     * Cell state, post tanh.
-     * Shape depends on data format (in layer config):<br>
-     * TNS -> [timeSteps, batchSize, numUnits]<br>
-     * NST -> [batchSize, numUnits, timeSteps]<br>
-     * NTS -> [batchSize, timeSteps, numUnits]<br>
+     * output at last step hL:
+     * [bS, nOut]   when directionMode FWD or BWD
+     * 2, bS, nOut] when directionMode  BIDIR_SUM, BIDIR_CONCAT or BIDIR_EXTRA_DIM
      */
-    private SDVariable h;
+    private SDVariable hL;
```

Comment:
Let's rename h, cL and hL to something more user friendly:
h -> timeSeriesOutput
hL -> lastTimeStepOutput
cL -> lastCellStateOutput